hoomd/array.py

```diff
@@ -0,0 +1,652 @@
+        out = kwargs.pop("out", None)
+        kwargs = {k: coerce_mock_to_array(v) for k, v in kwargs}
+        if out is not None:
+            if any(isinstance(o, HOOMDArray) for o in out):
+                kwargs['out'] = tuple((coerce_mock_to_array(v) for v in out))
+                getattr(ufunc, method)(*new_inputs, **kwargs)
+                return self
+        return getattr(ufunc, method)(*new_inputs, **kwargs)
+
+    def __getattr__(self, item):
+        """Used to mock `numpy.ndarray`'s interface."""
+        arr = self._coerce_to_ndarray()
+        return getattr(arr, item)
+
+    @property
+    def __array_interface__(self):
+        """Returns the information for a copy of the underlying data buffer.
+
+        This ensures that calls to functions like `numpy.array` do not cause
+        invalid access to the buffer. We must copy because ones we release
+        `numpy.ndarray` pointing to the underlying buffer we cannot guarentee
+        safety.
+        """
+        return np.array(self._coerce_to_ndarray(),
+                        copy=True).__array_interface__
+
+    def _coerce_to_ndarray(self):
+        """Provide a `numpy.ndarray` interface to the underlying buffer.
+
+        Raises a `HOOMDArrayError` when the provide callback returns False.
+        """
+        if self._callback():
+            if self._read_only:
+                arr = np.array(self._buffer, copy=False)
+                arr.flags['WRITEABLE'] = False
+                return arr
+            else:
+                return np.array(self._buffer, copy=False)
+        else:
+            raise HOOMDArrayError(
+                "Cannot access {} outside context manager. Use "
+                "numpy.array inside context manager instead.".format(
+                    self.__class__.__name__))
+
+    @property
+    def shape(self):
+        return self._coerce_to_ndarray().shape
+
+    @shape.setter
+    def shape(self, value):
+        raise HOOMDArrayError("Shape cannot be set on a {}. Use "
+                              "``array.reshape`` instead.".format(
+                                  self.__class__.__name__))
+
+    @property
+    def read_only(self):
+        return self._read_only
+
+    def __str__(self):
+        name = self.__class__.__name__
+        if self._callback():
+            return name + "(" + str(self._coerce_to_ndarray()) + ")"
+        else:
+            return name + "(INVALID)"
+
+    def __repr__(self):
+        name = self.__class__.__name__
+        if self._callback():
+            return name + "(" + str(self._coerce_to_ndarray()) + ")"
+        else:
+            return name + "(INVALID)"
+
+    def _repr_html_(self):
+        name = self.__class__.__name__
+        if self._callback():
+            return "<emph>" + name + "</emph>" \
+                + "(" + str(self._coerce_to_ndarray()) + ")"
+        else:
+            return "<emph>" + name + "</emph>" \
+                + "(<strong>INVALID</strong>)"
+
```

Comment:
This is nice, but also we don't do this anywhere else do we? I guess this is just to make it as numpy-array-like as possible?

Reply:
We don't it was just easy to add, and I could see potential for integrating some of jupyter's features like rich `repr`s with other classes.